quantum/process_keycode/process_auto_shift.c

```diff
@@ -79,24 +53,35 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
         return true;
     }
 
-    const uint16_t elapsed = TIMER_DIFF_16(record->event.time, autoshift_time);
-    autoshift_flush(record->event.time);
-
 #    ifndef AUTO_SHIFT_MODIFIERS
-    if (get_mods() ^ (autoshift_flags.holding_shift ? MOD_BIT(KC_LSFT) : 0)) {
+    if (get_mods() & (~MOD_BIT(KC_LSFT))) {
         return true;
     }
 #    endif
-#    ifndef TAPPING_FORCE_HOLD
-    if (elapsed < TAPPING_TERM && keycode == autoshift_lastkey && !autoshift_flags.lastshifted) {
-        // Allow a tap-then-hold to hold the unshifted key.
-        autoshift_lastkey = KC_NO;
-        return true;
+#    ifdef AUTO_SHIFT_REPEAT
+    const uint16_t elapsed = TIMER_DIFF_16(record->event.time, autoshift_time);
+#        ifndef AUTO_SHIFT_NO_AUTO_REPEAT
+    if (!autoshift_flags.lastshifted) {
+#        endif
+        if (elapsed < TAPPING_TERM && keycode == autoshift_lastkey) {
+            // Allow a tap-then-hold for keyrepeat.
+            if (!autoshift_flags.lastshifted) {
+                register_code(autoshift_lastkey);
```

Comment:
Would be good if we could return true here rather than registering the key ourselves.  Though you'd miss updating autoshift_time below if you returned directly.